central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,624 @@
+
+	s.queuePendingReports(ctx)
+	s.queueScheduledReports(ctx)
+	go s.runReports()
+	go s.runReports()
+}
+
+func (s *scheduler) Stop() {
+	s.stopper.Client().Stop()
+	err := s.stopper.Client().Stopped().Wait()
+	if err != nil {
+		log.Errorf("Error stopping vulnerability report scheduler : %v", err)
+	}
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
+			reportRequest := s.selectNextRunnableReport()
+			if reportRequest == nil {
+				continue
+			}
+			if err := s.concurrencySema.Acquire(scheduledCtx, 1); err != nil {
+				log.Errorf("Error acquiring semaphore to run new report: %v", err)
+				continue
+			}
+			log.Infof("Executing report '%s' at %v", reportRequest.ReportConfig.GetName(), time.Now().Format(time.RFC822))
+			go s.runSingleReport(reportRequest)
+		}
+	}
+}
+
+func (s *scheduler) selectNextRunnableReport() *ReportRequest {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	reportIdx := sliceutils.FindMatching(s.reportsQueue, func(req *ReportRequest) bool {
+		return !s.runningReportConfigs.Contains(req.ReportConfig.GetId())
+	})
+
+	if reportIdx < 0 {
+		// did not find any reports to run
+		return nil
+	}
+	request := s.reportsQueue[reportIdx]
+	s.reportsQueue = slices.Delete(s.reportsQueue, reportIdx, reportIdx+1)
+	s.runningReportConfigs.Add(request.ReportConfig.GetId())
+	return request
+}
+
+func (s *scheduler) runSingleReport(req *ReportRequest) {
+	defer s.notifySchedulerForMoreReports()
+	defer s.concurrencySema.Release(1)
+	defer s.removeFromRunningReportConfigs(req.ReportConfig.GetId())
+
+	var err error
+	if req.ReportConfig.GetVulnReportFilters().GetSinceLastSentScheduledReport() {
+		req.dataStartTime, err = s.lastSuccesfulScheduledReportTime(req.Ctx, req.ReportConfig)
+		if err != nil {
+			s.logAndUpsertError(errors.Wrap(err, "Error finding last successful scheduled report time"), req)
+			return
+		}
+	} else if req.ReportConfig.GetVulnReportFilters().GetSinceStartDate() != nil {
+		req.dataStartTime = req.ReportConfig.GetVulnReportFilters().GetSinceStartDate()
+	}
+
+	// Change report status to PREPARING
+	err = s.upsertReportStatus(req.Ctx, req.ReportMetadata, storage.ReportStatus_PREPARING)
+	if err != nil {
+		s.logAndUpsertError(errors.Wrap(err, "Error changing report status to PREPARING"), req)
+		return
+	}
+
+	err = s.generateAndSendReportResult(req)
+	if err != nil {
+		s.logAndUpsertError(err, req)
+		return
+	}
+
+	// Change report status to SUCCESS
+	req.ReportMetadata.ReportStatus.CompletedAt = types.TimestampNow()
+	err = s.upsertReportStatus(req.Ctx, req.ReportMetadata, storage.ReportStatus_SUCCESS)
+	if err != nil {
+		s.logAndUpsertError(errors.Wrap(err, "Error changing report status to SUCCESS"), req)
+		return
+	}
+	s.takeReportSnapshot(req)
+}
+
+func (s *scheduler) notifySchedulerForMoreReports() {
+	s.sendNewReports <- true
```

Comment:
Added checks to make sure `Start` is only called once per instance of scheduler. If it is called more than once, it will log an error and return. Made same changes for `Stop`.